### ISSUE TITLE
Supplementary_Table_S6.py

### ISSUE THREAD

**Issue by Curiosity-Mars** [Curiosity-Mars/rarp-psa-trajectory-reproducibility#14]
"""Supplementary Table S6. Figure 3 model specification and performance.

Panels A-E document the cohort and modeling specification, predictor
missingness, out-of-fold performance, paired AUC comparisons, and prespecified
sensitivity analyses.

Prerequisites: run 00_Common_Setup.py, Table 2, Figure 2, and Figure 3.
"""

from __future__ import annotations

import numpy as np
import pandas as pd
from sklearn.metrics import average_precision_score, balanced_accuracy_score

if "OUTPUT_DIR" not in globals():
    raise RuntimeError("Run 00_Common_Setup.py before this script")

LINEAR_FILE = OUTPUT_DIR / "Table2" / "Table2_Per_Patient_Linear_Descriptors.csv"
NONLINEAR_FILE = OUTPUT_DIR / "Figure2" / "Figure2_Per_Patient_Nonlinear_Indicators.csv"
FIGURE3_DIR = OUTPUT_DIR / "Figure3"
OOF_FILE = FIGURE3_DIR / "Figure3_OOF_Predictions.csv"
PERFORMANCE_FILE = FIGURE3_DIR / "Figure3_Model_Performance.csv"
COMPARISON_FILE = FIGURE3_DIR / "Figure3_Paired_AUC_Comparisons.csv"
SENSITIVITY_FILE = FIGURE3_DIR / "Figure3_Sensitivity_Analyses.csv"
for path in (LINEAR_FILE, NONLINEAR_FILE, OOF_FILE, PERFORMANCE_FILE, COMPARISON_FILE, SENSITIVITY_FILE):
    if not path.exists():
        raise FileNotFoundError(f"Missing prerequisite output: {path}")

OUTDIR = OUTPUT_DIR / "Supplementary_Table_S6"
OUTDIR.mkdir(parents=True, exist_ok=True)

linear = pd.read_csv(LINEAR_FILE)
nonlinear = pd.read_csv(NONLINEAR_FILE)
model_data = linear.merge(nonlinear, on=["Analysis_Row_ID", "Cluster"], validate="one_to_one")
oof = pd.read_csv(OOF_FILE)
performance = pd.read_csv(PERFORMANCE_FILE)
comparisons = pd.read_csv(COMPARISON_FILE)
sensitivity = pd.read_csv(SENSITIVITY_FILE)

panel_a = pd.DataFrame(
    [
        ("Source cohort", "1,685"),
        ("Analysis-ready dataset", "1,455"),
        ("Figure 2 trajectory cohort", "1,434"),
        ("Reference C0 / C1 / C2", "78 / 984 / 372"),
        ("Classification target", "C1 vs C0+C2"),
        ("Cross-validation", "Stratified 5-fold, same folds for all models"),
        ("Preprocessing", "Median imputation and standardization inside each training fold"),
        ("Primary uncertainty", "Patient-level percentile bootstrap of pooled OOF predictions"),
        ("Primary nonlinear features", "operational t*, max a(t), mean a(t)"),
        ("Indicator evaluability", "Sensitivity variable; not a primary predictive feature"),
    ],
    columns=["Assessment item", "Value"],
)

missing_features = {
    "overall_slope": "overall_raw_slope",
    "log_slope": "overall_log_slope",
    "slope_0_24": "raw_slope_POM0_24",
    "slope_24_60": "raw_slope_POM24_60",
    "slope_60_120": "raw_slope_POM60_120",
    "t_star": "operational_t_star",
    "max_at": "max_a_t",
    "mean_at": "mean_a_t",
    "indicator_evaluable": "indicator_evaluable",
    "t_star_boundary": "t_star_boundary",
}
panel_b_rows = []
for display, column in missing_features.items():
    missing_n = int(model_data[column].isna().sum())
    panel_b_rows.append(
        {"Feature": display, "Missing, n": missing_n, "Missing, %": round(100 * missing_n / len(model_data), 1)}
    )
panel_b = pd.DataFrame(panel_b_rows)

y = oof["target_cluster1"].to_numpy(int)
panel_c_rows = []
for model in ("Linear", "Nonlinear", "Integrative"):
    record = performance.loc[performance["Model"] == model].iloc[0]
    probability = oof[f"probability_{model}"].to_numpy(float)
    panel_c_rows.append(
        {
            "Model": model,
            "N": len(y),
            "ROC AUC": f"{record['ROC_AUC']:.3f}",
            "95% CI": f"{record['CI_low']:.3f}-{record['CI_high']:.3f}",
            "PR AUC": f"{average_precision_score(y, probability):.3f}",
            "Balanced accuracy*": f"{balanced_accuracy_score(y, probability >= 0.5):.3f}",
        }
    )
panel_c = pd.DataFrame(panel_c_rows)

def signed(value: float) -> str:
    return f"{value:+.3f}"

panel_d_rows = []
for label, source_label in (
    ("Nonlinear - Linear", "Nonlinear − Linear"),
    ("Integrative - Linear", "Integrative − Linear"),
    ("Integrative - Nonlinear", "Integrative − Nonlinear"),
):
    record = comparisons.loc[comparisons["Comparison"] == source_label].iloc[0]
    p_value = float(record["P_bootstrap_two_sided"])
    panel_d_rows.append(
        {
            "Comparison": label,
            "Delta AUC": signed(float(record["Delta_AUC"])),
            "95% CI": f"{signed(float(record['CI_low']))} to {signed(float(record['CI_high']))}",
            "Two-sided bootstrap P": "<0.001" if p_value < 0.001 else f"{p_value:.2f}",
        }
    )
panel_d = pd.DataFrame(panel_d_rows)

panel_e_rows = []
for analysis_name, source_name in (
    ("Evaluable only", "Evaluable only"),
    ("Boundary t* as missing", "Boundary t* treated as missing"),
):
    for model in ("Linear", "Nonlinear", "Integrative"):
        record = sensitivity.loc[
            (sensitivity["Analysis"] == source_name) & (sensitivity["Model"] == model)
        ].iloc[0]
        panel_e_rows.append(
            {
                "Analysis": analysis_name,
                "Model": model,
                "N": int(record["N"]),
                "ROC AUC": f"{record['ROC_AUC']:.3f}",
                "95% CI": f"{record['CI_low']:.3f}-{record['CI_high']:.3f}",
            }
        )
panel_e = pd.DataFrame(panel_e_rows)

expected_b = [["overall_slope", 0, 0.0], ["log_slope", 0, 0.0], ["slope_0_24", 0, 0.0], ["slope_24_60", 398, 27.8], ["slope_60_120", 969, 67.6], ["t_star", 22, 1.5], ["max_at", 22, 1.5], ["mean_at", 22, 1.5], ["indicator_evaluable", 0, 0.0], ["t_star_boundary", 22, 1.5]]
expected_c = [["Linear", 1434, "0.640", "0.612-0.667", "0.824", "0.487"], ["Nonlinear", 1434, "0.735", "0.708-0.759", "0.880", "0.517"], ["Integrative", 1434, "0.734", "0.708-0.758", "0.874", "0.525"]]
expected_d = [["Nonlinear - Linear", "+0.095", "+0.069 to +0.121", "<0.001"], ["Integrative - Linear", "+0.094", "+0.069 to +0.119", "<0.001"], ["Integrative - Nonlinear", "-0.001", "-0.008 to +0.004", "0.74"]]
expected_e = [["Evaluable only", "Linear", 1412, "0.635", "0.607-0.662"], ["Evaluable only", "Nonlinear", 1412, "0.736", "0.710-0.760"], ["Evaluable only", "Integrative", 1412, "0.759", "0.733-0.784"], ["Boundary t* as missing", "Linear", 1434, "0.640", "0.612-0.667"], ["Boundary t* as missing", "Nonlinear", 1434, "0.735", "0.709-0.758"], ["Boundary t* as missing", "Integrative", 1434, "0.734", "0.708-0.758"]]
for label, observed, expected in (
    ("B", panel_b.values.tolist(), expected_b),
    ("C", panel_c.values.tolist(), expected_c),
    ("D", panel_d.values.tolist(), expected_d),
    ("E", panel_e.values.tolist(), expected_e),
):
    if observed != expected:
        raise AssertionError(f"Supplementary Table S6 panel {label} mismatch: {observed}")

panels = {"A": panel_a, "B": panel_b, "C": panel_c, "D": panel_d, "E": panel_e}
for label, frame in panels.items():
    frame.to_csv(OUTDIR / f"Supplementary_Table_S6{label}.csv", index=False, encoding="utf-8-sig")
titles = {"A": "Cohort and modeling specification", "B": "Predictor missingness", "C": "Model performance", "D": "Paired bootstrap AUC comparisons", "E": "Sensitivity analyses"}
parts = ["# Supplementary Table S6. Figure 3 predictive-model specification and classification performance"]
for label, frame in panels.items():
    parts.extend([f"## {label}. {titles[label]}", frame.to_markdown(index=False)])
(OUTDIR / "Supplementary_Table_S6.md").write_text("\n\n".join(parts) + "\n", encoding="utf-8")

print("SUPPLEMENTARY TABLE S6 COMPLETION GATE PASSED")
for label, frame in panels.items():
    print(f"\nPanel {label}\n{frame.to_string(index=False)}")
